### ISSUE TITLE
Support Quake 2 & 3 BSP files

### ISSUE THREAD

**Issue by krazyjakee** [jitspoe/godot_bsp_importer#4]
There are many BSP formats but a good starting point would be Q2 and Q3.

Some open questions:

Where are some resources for learning about the differences between the formats?

Is this outside the scope of this project?

Other possible BSP formats https://github.com/zturtleman/bsp-sekai

**Comment by jitspoe** [comment]
Definitely within the scope of this project!  I've vaguely started supporting other formats but didn't get very far as it wasn't a priority for my project.

One thing to be careful of is that most of the tools and engines out there that read and write BSP files are under the GPL license, so copying code from them isn't compatible with the MIT license.  There are some resources floating around explaining the formats that were from reverse engineering the format (like before the source was released).

For example, this info about the Quake 2 format came out before the source code was released: https://www.flipcode.com/archives/Quake_2_BSP_File_Format.shtml

**Comment by Skaruts** [comment]
Is it ok to use Valve map format in TB? It seems like qbsp has no problems with it, and my test maps seem to work fine as well. Valve format supports better texture projections, so you can transfer textures between faces much better with it, and it doesn't create brushes with some textures inverted as the standard format does (noticeable with textures that have text in them, like triggers, clips, etc).


It seems qbsp limits filenames to 15 characters, though, and I don't know if that can be changed.
```
WARNING: texture DiamondPlate008C name too long for Quake miptex
```

**Comment by jitspoe** [comment]
Yep.  That only impacts the .map file, so if you have a modern qbsp compiler that can handle the valve format, it's fine.

And, yeah, the texture names are pretty limited, which is a down side.  I think the Q2 format supports longer texture names, but q2bsp support would have to be implemented, which might take a while.

If you wanted to have more properly named materials, you could use longer names for the materials, shortened names for the quake texture names, and use the material remap dictionary on the importer, but that's kind of yucky.  In this case, you could strip off a 0 or two.  Probably not going to have more than 99 diamond plates. :D

**Comment by CSLRDoesntGameDev** [comment]
this is a year too late probably, anyway https://github.com/jitspoe/godot_bsp_importer/pull/24

**Comment by daniel-abbott** [comment]
> this is a year too late probably, anyway [#24](https://github.com/jitspoe/godot_bsp_importer/pull/24)

Never too late, Quake 2's format _may_ have some practical advantages, at least with full texture paths. Potentially worth bumping this into a new issue, however it seems that it doesn't actually import Q2 entities? At least not lights... haven't had time to actually dig into why yet.

**Comment by jitspoe** [comment]
What's currently merged is just a partial implementation with no entity support.  I think CSLR's fork has some more functionality that hasn't been pulled over yet.

**Comment by CSLRDoesntGameDev** [comment]
> I think CSLR's fork has some more functionality that hasn't been pulled over yet.

Yeah, there's some quake 3 stuff but i'm not too focused on that right now, been working on some general refactors since my code was a flaming dumpster fire.

as per what @daniel-abbott said, i do believe the most recent update (#32) contained the move over from a entity reader i wrote over to what was pre-existing, is it possible that quake 2's light entity is formatted differently?

**Comment by CSLRDoesntGameDev** [comment]
@daniel-abbott i found the issue causing the quake 2 entities not being imported, working on a fix for it now. its reading the entities in but not adding them as children of the packed scene. thanks for catching that :)

**Comment by CSLRDoesntGameDev** [comment]
Been Working on the Quake 3 Side while improving the Quake 2 code a tad bit, thought you might want to see lizard person from OA-KOTH1 @krazyjakee   

![Image](https://github.com/user-attachments/assets/0e34849e-e145-468d-a382-4259e2e802ab)

**Comment by krazyjakee** [comment]
@CSLRDoesntGameDev it's not even my birthday

**Comment by jitspoe** [comment]
CSLR's stuff is integrated in, now, but not thoroughly tested.

**Comment by jitspoe** [comment]
I'm going to close this now.  If there are any specific issues with Q2 or Q3, feel free to open new issues for those.
